crates/fuel-core/src/database.rs

```diff
@@ -83,164 +82,89 @@ pub mod message;
 pub mod metadata;
 pub mod sealed_block;
 pub mod state;
-pub mod statistic;
 pub mod storage;
-pub mod transaction;
 pub mod transactions;
 
 #[derive(Clone, Debug)]
 pub struct Database<Description = OnChain>
 where
     Description: DatabaseDescription,
 {
-    data: StructuredStorage<DataSource<Description>>,
-    // used for RAII
-    _drop: Arc<DropResources>,
-}
-
-type DropFn = Box<dyn FnOnce() + Send + Sync>;
-#[derive(Default)]
-struct DropResources {
-    // move resources into this closure to have them dropped when db drops
-    drop: Option<DropFn>,
-}
-
-impl fmt::Debug for DropResources {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "DropResources")
-    }
-}
-
-impl<F: 'static + FnOnce() + Send + Sync> From<F> for DropResources {
-    fn from(closure: F) -> Self {
-        Self {
-            drop: Option::Some(Box::new(closure)),
-        }
-    }
-}
-
-impl Drop for DropResources {
-    fn drop(&mut self) {
-        if let Some(drop) = self.drop.take() {
-            (drop)()
-        }
-    }
+    height: Arc<Mutex<Option<Description::Height>>>,
+    data: DataSource<Description>,
 }
 
 impl<Description> Database<Description>
 where
     Description: DatabaseDescription,
+    Self: StorageInspect<MetadataTable<Description>, Error = StorageError>,
 {
-    pub fn new<D>(data_source: D) -> Self
-    where
-        D: Into<DataSource<Description>>,
-    {
-        Self {
-            data: StructuredStorage::new(data_source.into()),
-            _drop: Default::default(),
-        }
-    }
+    pub fn new(data_source: DataSource<Description>) -> Self {
+        let mut database = Self {
+            height: Arc::new(Mutex::new(None)),
+            data: data_source,
+        };
+        let height = database
+            .latest_height()
+            .expect("Failed to get latest height during creation of the database");
+
+        database.height = Arc::new(Mutex::new(height));
```

Comment:
Why do we need to set the `height` if there is a `latest_height()` method that works without that field?

Reply:
I see the TODO to add this below. Is this just to reduce lookups?

Reply:
It reduces lookup into the database, plus avoids any concurrent modification of the database because we verify the block height during `commit_changes`, and it should advance each commit.